warehouse/oidc/services.py

```diff
@@ -0,0 +1,172 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+
+import redis
+import requests
+import sentry_sdk
+
+from jwt import PyJWK
+from zope.interface import implementer
+
+from warehouse.metrics.interfaces import IMetricsService
+from warehouse.oidc.interfaces import IOIDCProviderService
+from warehouse.utils import oidc
+
+
+@implementer(IOIDCProviderService)
+class OIDCProviderService:
+    def __init__(self, provider, cache_url, metrics):
+        self.provider = provider
+        self.cache_url = cache_url
+        self.metrics = metrics
+
+        self._issuer_url = oidc.OIDC_PROVIDERS[self.provider]
```

Comment:
@di Is this what you had in mind, re: https://github.com/pypa/warehouse/pull/10628/files#r794061431?

Reply:
No, see https://github.com/pypa/warehouse/pull/10628#discussion_r794063016 and https://github.com/pypa/warehouse/pull/10628#discussion_r794061762 and https://github.com/pypa/warehouse/pull/10628#discussion_r794088544